.evergreen/socks5srv.py

```diff
@@ -0,0 +1,237 @@
+      except OSError as msg:
+        outgoing.close()
+        continue
+      break
+    return outgoing
+
+  def handle(self):
+    """Handle the Socks5 communication with a freshly connected client"""
+
+    # Client greeting
+    if self.request.recv(1) != b'\x05': # Socks5 only
+      return
+    n_auth = self.request.recv(1)
+    client_auth_methods = self.read_exact(n_auth)
+    if client_auth_methods is None:
+      return
+
+    # choose either no-auth or username/password
+    required_auth_method = b'\x00' if self.server.args.auth is None else b'\x02'
+    if required_auth_method not in client_auth_methods:
+      self.request.sendall(b'\x05\xff')
+      return
+
+    self.request.sendall(b'\x05' + required_auth_method)
+    if required_auth_method == b'\x02':
+      auth_version = self.request.recv(1)
+      if auth_version != b'\x01': # Only username/password auth v1
+        return
+      username_len = self.request.recv(1)
+      username = self.read_exact(username_len)
+      password_len = self.request.recv(1)
+      password = self.read_exact(password_len)
+      if username is None or password is None:
+        return
+      if username.decode('utf8') + ':' + password.decode('utf8') != self.server.args.auth:
+        return
+      self.request.sendall(b'\x01\x00') # auth success
+
+    if self.request.recv(1) != b'\x05': # Socks5 only
+      return
+    if self.request.recv(1) != b'\x01': # Outgoing TCP only
+      return
+    if self.request.recv(1) != b'\x00': # Reserved, must be 0
+      return
+
+    addrtype = self.request.recv(1)
+    dst = None
+    if addrtype == b'\x01': # IPv4
+      ipv4raw = self.read_exact(4)
+      if ipv4raw is not None:
+        dst = '.'.join(['{}'] * 4).format(*ipv4raw)
+    elif addrtype == b'\x03': # Domain
+      domain_len = self.request.recv(1)
+      dst = self.read_exact(domain_len)
+    elif addrtype == b'\x04': # IPv6
+      ipv6raw = self.read_exact(16)
+      if ipv6raw is not None:
+        dst = ':'.join(['{:0>2x}{:0>2x}'] * 8).format(*ipv6raw)
+    else:
+      return
+
+    if dst is None:
+      return
+
+    portraw = self.read_exact(2)
+    port = portraw[0] * 256 + portraw[1]
+
+    (dst, port) = self.server.address_remapper.remap((dst, port))
+
+    outgoing = self.create_outgoing_tcp_connection(dst, port)
+    if outgoing is None:
+      self.request.sendall(b'\x05\x01\x00') # just report a general failure
+      return
+    # success response, do not bother actually stating the locally bound
+    # host/port address and instead always say 127.0.0.1:4096.
+    # for our use case, the client will not be making meaningful use
+    # of this anyway
+    self.request.sendall(b'\x05\x00\x00\x01\x7f\x00\x00\x01\x10\x00')
+
+    self.raw_proxy(self.request, outgoing)
+
+  def raw_proxy(self, a, b):
+    """Proxy data between sockets a and b as-is"""
+
+    with a, b:
+      while True:
+        try:
+          (readable, _, _) = select.select([a, b], [], [])
```

Comment:
select will fail if the fd is > 1024 so we should prefer [poll](https://docs.python.org/3/library/select.html#poll-objects) but since this is test only it's not a big deal. We're not planning to test this with >1000 connections are we?

Reply:
Also here, I agree that this is just fine for our usage – we are not planning to run this with more than 1000 connections at a time for sure :)